moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.kt

```diff
@@ -0,0 +1,1169 @@
+        if (peekStack == JsonScope.NONEMPTY_OBJECT) {
+          val c = nextNonWhitespace(true).toChar()
+          buffer.readByte() // Consume '}' or ','.
+          when (c) {
+            '}' -> return PEEKED_END_OBJECT.also { peeked = it }
+            ',' -> Unit /*no op*/
+            ';' -> checkLenient()
+            else -> throw syntaxError("Unterminated object")
+          }
+        }
+        return when (val c = nextNonWhitespace(true).toChar()) {
+          '"' -> {
+            buffer.readByte() // consume the '\"'.
+            PEEKED_DOUBLE_QUOTED_NAME.also { peeked = it }
+          }
+          '\'' -> {
+            buffer.readByte() // consume the '\''.
+            checkLenient()
+            PEEKED_SINGLE_QUOTED_NAME.also { peeked = it }
+          }
+          '}' -> if (peekStack != JsonScope.NONEMPTY_OBJECT) {
+            buffer.readByte() // consume the '}'.
+            PEEKED_END_OBJECT.also { peeked = it }
+          } else {
+            throw syntaxError("Expected name")
+          }
+          else -> {
+            checkLenient()
+            if (isLiteral(c.code)) {
+              PEEKED_UNQUOTED_NAME.also { peeked = it }
+            } else {
+              throw syntaxError("Expected name")
+            }
+          }
+        }
+      }
+      JsonScope.DANGLING_NAME -> {
+        scopes[stackSize - 1] = JsonScope.NONEMPTY_OBJECT
+        // Look for a colon before the value.
+        val c = nextNonWhitespace(true).toChar()
+        buffer.readByte() // Consume ':'.
+        when (c) {
+          ':' -> Unit /*no op*/
+          '=' -> {
+            checkLenient()
+            if (source.request(1) && buffer[0] == '>'.code.toByte()) {
+              buffer.readByte() // Consume '>'.
+            }
+          }
+          else -> throw syntaxError("Expected ':'")
+        }
+      }
+      JsonScope.EMPTY_DOCUMENT -> scopes[stackSize - 1] = JsonScope.NONEMPTY_DOCUMENT
+      JsonScope.NONEMPTY_DOCUMENT -> {
+        if (nextNonWhitespace(false) == -1) {
+          return PEEKED_EOF.also { peeked = it }
+        } else {
+          checkLenient()
+        }
+      }
+      JsonScope.STREAMING_VALUE -> {
+        valueSource!!.discard()
+        valueSource = null
+        stackSize--
+        return doPeek()
+      }
+      else -> check(peekStack != JsonScope.CLOSED) { "JsonReader is closed" }
+    }
+    // "fallthrough from previous `when`
+    when (nextNonWhitespace(true).toChar()) {
+      ']' -> {
+        if (peekStack == JsonScope.EMPTY_ARRAY) {
+          buffer.readByte() // Consume ']'.
+          return PEEKED_END_ARRAY.also { peeked = it }
+        }
+        // In lenient mode, a 0-length literal in an array means 'null'.
+        return when (peekStack) {
+          JsonScope.EMPTY_ARRAY, JsonScope.NONEMPTY_ARRAY -> {
+            checkLenient()
+            PEEKED_NULL.also { peeked = it }
+          }
+          else -> throw syntaxError("Unexpected value")
```

Comment:
in the original java sources, this code block was a fallthrough for the `';', ','` case; I figured the correct course of action is to duplicate the logic?